pennylane/beta/queuing/circuit.py

```diff
@@ -0,0 +1,499 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the Circuit class which is used to represent the internal
+quantum datastructure of a QNode.
+"""
+from collections import OrderedDict
+
+from pennylane.operation import Operation, Expectation, Observable, Probability, Sample, Variance
+from .queuing import AnnotatedQueue, QueuingContext
+from .measure import MeasurementProcess
+
+
+def _decompose_ops(ops, device=None):
+    """Recursively loop through a list and decompose
+    operations that are not supported by a device.
+
+    If a device is not provided, the decomposition continues until
+    an operation cannot be decomposed.
+
+    Args:
+        ops (List[~.Operation]): operation queue
+        device (~.Device): a PennyLane device
+    """
+    new_ops = []
+
+    for op in ops:
+        if device is not None and device.supports_operation(op.name):
+            new_ops.append(op)
+        else:
+            try:
+                decomposed_ops = op.decomposition(*op.data, wires=op.wires)
+                decomposition = _decompose_queue(decomposed_ops, device)
+                new_ops.extend(decomposition)
+            except NotImplementedError:
+                new_ops.append(op)
+                break
+
+    return new_ops
+
+
+class Circuit(AnnotatedQueue):
+    def __init__(self, name=None):
+        self.name = name
+        super().__init__()
+
+    def __enter__(self):
+        if isinstance(QueuingContext.active_context(), Circuit):
+            QueuingContext.append(self, name=self.name)

```

Comment:
So this will nest itself as a subcircuit of any active circuit? :+1:

Reply:
Yep!